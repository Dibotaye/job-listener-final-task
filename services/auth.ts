import type { SignupFormData, SigninFormData, VerifyEmailData, AuthResponse } from "../types/auth"

const BASE_URL = "https://akil-backend.onrender.com"

/**
 * Sign up a new user
 * @param userData - User signup data
 * @returns Promise<AuthResponse> Response from signup API
 */
export async function signupUser(userData: SignupFormData): Promise<AuthResponse> {
  try {
    const response = await fetch(`${BASE_URL}/signup`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(userData),
    })

    const data: AuthResponse = await response.json()

    if (!response.ok) {
      throw new Error(data.message || `HTTP error! status: ${response.status}`)
    }

    return data
  } catch (error) {
    console.error("Signup error:", error)

    if (error instanceof TypeError && error.message.includes("fetch")) {
      throw new Error("Network error: Please check your internet connection")
    }

    throw error instanceof Error ? error : new Error("Signup failed")
  }
}

/**
 * Sign in an existing user
 * @param credentials - User signin credentials
 * @returns Promise<AuthResponse> Response from signin API with access token
 */
export async function signinUser(credentials: SigninFormData): Promise<AuthResponse> {
  try {
    const response = await fetch(`${BASE_URL}/login`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(credentials),
    })

    const data: AuthResponse = await response.json()

    if (!response.ok) {
      // Handle specific error cases
      if (response.status === 401) {
        throw new Error("Invalid email or password")
      }
      throw new Error(data.message || `HTTP error! status: ${response.status}`)
    }

    if (!data.success) {
      throw new Error(data.message || "Login failed")
    }

    return data
  } catch (error) {
    console.error("Signin error:", error)

    if (error instanceof TypeError && error.message.includes("fetch")) {
      throw new Error("Network error: Please check your internet connection")
    }

    throw error instanceof Error ? error : new Error("Login failed")
  }
}

/**
 * Verify user email with OTP
 * @param verificationData - Email and OTP data
 * @returns Promise<AuthResponse> Response from verification API
 */
export async function verifyEmail(verificationData: VerifyEmailData): Promise<AuthResponse> {
  try {
    const response = await fetch(`${BASE_URL}/verify-email`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(verificationData),
    })

    const data: AuthResponse = await response.json()

    if (!response.ok) {
      if (response.status === 400) {
        throw new Error("Invalid verification code")
      }
      throw new Error(data.message || `HTTP error! status: ${response.status}`)
    }

    if (!data.success) {
      throw new Error(data.message || "Email verification failed")
    }

    return data
  } catch (error) {
    console.error("Email verification error:", error)

    if (error instanceof TypeError && error.message.includes("fetch")) {
      throw new Error("Network error: Please check your internet connection")
    }

    throw error instanceof Error ? error : new Error("Email verification failed")
  }
}

/**
 * Get stored access token
 * @returns string | null Access token from localStorage
 */
export function getAccessToken(): string | null {
  if (typeof window !== "undefined") {
    return localStorage.getItem("accessToken")
  }
  return null
}

/**
 * Get stored user data
 * @returns any | null User data from localStorage
 */
export function getUserData(): any | null {
  if (typeof window !== "undefined") {
    const userData = localStorage.getItem("userData")
    return userData ? JSON.parse(userData) : null
  }
  return null
}

/**
 * Check if user is authenticated
 * @returns boolean Whether user has valid access token
 */
export function isAuthenticated(): boolean {
  return !!getAccessToken()
}

/**
 * Logout user by clearing stored data
 */
export function logout(): void {
  if (typeof window !== "undefined") {
    localStorage.removeItem("accessToken")
    localStorage.removeItem("refreshToken")
    localStorage.removeItem("userData")
    sessionStorage.removeItem("verificationEmail")
  }
}
