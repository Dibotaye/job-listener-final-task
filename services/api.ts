import type { Job, ApiResponse } from "../types/job"
import { getAccessToken } from "./auth"

const BASE_URL = "https://akil-backend.onrender.com"

/**
 * Fetches all opportunities from the API
 * @returns Promise<Job[]> Array of job opportunities
 */
export async function fetchOpportunities(): Promise<Job[]> {
  try {
    const token = getAccessToken()
    const headers: HeadersInit = {
      "Content-Type": "application/json",
    }

    // Add authorization header if token exists
    if (token) {
      headers["Authorization"] = `Bearer ${token}`
    }

    const response = await fetch(`${BASE_URL}/opportunities/search`, {
      method: "GET",
      headers,
      cache: "no-store",
    })

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    const data: ApiResponse = await response.json()

    if (!data.success) {
      throw new Error(data.message || "Failed to fetch opportunities")
    }

    return data.data || []
  } catch (error) {
    console.error("Error fetching opportunities:", error)

    if (error instanceof TypeError && error.message.includes("fetch")) {
      throw new Error("Network error: Please check your internet connection")
    }

    throw error instanceof Error ? error : new Error("Unknown error occurred")
  }
}

/**
 * Search opportunities by title
 * @param query - Search query
 * @returns Promise<Job[]> Array of matching job opportunities
 */
export async function searchOpportunities(query: string): Promise<Job[]> {
  try {
    const token = getAccessToken()
    const headers: HeadersInit = {
      "Content-Type": "application/json",
    }

    if (token) {
      headers["Authorization"] = `Bearer ${token}`
    }

    const response = await fetch(`${BASE_URL}/opportunities/search?q=${encodeURIComponent(query)}`, {
      method: "GET",
      headers,
      cache: "no-store",
    })

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    const data: ApiResponse = await response.json()

    if (!data.success) {
      throw new Error(data.message || "Failed to search opportunities")
    }

    // Filter results by title if API doesn't support title search
    const jobs = data.data || []
    return jobs.filter((job) => job.title.toLowerCase().includes(query.toLowerCase()))
  } catch (error) {
    console.error("Error searching opportunities:", error)

    if (error instanceof TypeError && error.message.includes("fetch")) {
      throw new Error("Network error: Please check your internet connection")
    }

    throw error instanceof Error ? error : new Error("Search failed")
  }
}

/**
 * Fetches a specific opportunity by ID
 * @param id - The opportunity ID
 * @returns Promise<Job> The job opportunity details
 */
export async function fetchOpportunityById(id: string): Promise<Job> {
  try {
    if (!id) {
      throw new Error("Job ID is required")
    }

    const token = getAccessToken()
    const headers: HeadersInit = {
      "Content-Type": "application/json",
    }

    if (token) {
      headers["Authorization"] = `Bearer ${token}`
    }

    const response = await fetch(`${BASE_URL}/opportunities/${id}`, {
      method: "GET",
      headers,
      cache: "no-store",
    })

    if (!response.ok) {
      if (response.status === 404) {
        throw new Error("Job opportunity not found")
      }
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    const data: ApiResponse = await response.json()

    if (!data.success) {
      throw new Error(data.message || "Failed to fetch job details")
    }

    if (!data.data) {
      throw new Error("No job data received")
    }

    return data.data
  } catch (error) {
    console.error("Error fetching opportunity by ID:", error)

    if (error instanceof TypeError && error.message.includes("fetch")) {
      throw new Error("Network error: Please check your internet connection")
    }

    throw error instanceof Error ? error : new Error("Unknown error occurred")
  }
}
