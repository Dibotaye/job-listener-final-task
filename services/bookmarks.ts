import { getAccessToken } from "./auth"

const BASE_URL = "https://akil-backend.onrender.com"

export interface Bookmark {
  eventID: string
  dateBookmarked: string
  datePosted: string
  logoUrl: string
  opType: string
  orgName: string
  title: string
  location: string[]
}

export interface BookmarkResponse {
  success: boolean
  message: string
  data?: Bookmark[]
  errors?: any
}


export async function getBookmarks(): Promise<Bookmark[]> {
  try {
    const token = getAccessToken()
    if (!token) {
      throw new Error("Authentication required")
    }

    const response = await fetch(`${BASE_URL}/bookmarks`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
    })

    if (!response.ok) {
      if (response.status === 401) {
        throw new Error("Authentication expired. Please login again.")
      }
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    const data: BookmarkResponse = await response.json()

    if (!data.success) {
      throw new Error(data.message || "Failed to fetch bookmarks")
    }

    return data.data || []
  } catch (error) {
    console.error("Error fetching bookmarks:", error)

    if (error instanceof TypeError && error.message.includes("fetch")) {
      throw new Error("Network error: Please check your internet connection")
    }

    throw error instanceof Error ? error : new Error("Failed to fetch bookmarks")
  }
}


export async function addBookmark(jobId: string): Promise<boolean> {
  try {
    const token = getAccessToken()
    if (!token) {
      throw new Error("Authentication required")
    }

    const response = await fetch(`${BASE_URL}/bookmarks/${jobId}`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify({}), 
    })

    if (!response.ok) {
      if (response.status === 401) {
        throw new Error("Authentication expired. Please login again.")
      }
      if (response.status === 409) {
        throw new Error("Job is already bookmarked")
      }
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    const data: BookmarkResponse = await response.json()
    return data.success
  } catch (error) {
    console.error("Error adding bookmark:", error)

    if (error instanceof TypeError && error.message.includes("fetch")) {
      throw new Error("Network error: Please check your internet connection")
    }

    throw error instanceof Error ? error : new Error("Failed to bookmark job")
  }
}


export async function removeBookmark(jobId: string): Promise<boolean> {
  try {
    const token = getAccessToken()
    if (!token) {
      throw new Error("Authentication required")
    }

    const response = await fetch(`${BASE_URL}/bookmarks/${jobId}`, {
      method: "DELETE",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
    })

    if (!response.ok) {
      if (response.status === 401) {
        throw new Error("Authentication expired. Please login again.")
      }
      if (response.status === 404) {
        throw new Error("Bookmark not found")
      }
      throw new Error(`HTTP error! status: ${response.status}`)
    }

    const data: BookmarkResponse = await response.json()
    return data.success
  } catch (error) {
    console.error("Error removing bookmark:", error)

    if (error instanceof TypeError && error.message.includes("fetch")) {
      throw new Error("Network error: Please check your internet connection")
    }

    throw error instanceof Error ? error : new Error("Failed to remove bookmark")
  }
}


export async function toggleBookmark(jobId: string, currentStatus: boolean): Promise<boolean> {
  try {
    if (currentStatus) {
      await removeBookmark(jobId)
      return false
    } else {
      await addBookmark(jobId)
      return true
    }
  } catch (error) {
    throw error instanceof Error ? error : new Error("Failed to toggle bookmark")
  }
}
