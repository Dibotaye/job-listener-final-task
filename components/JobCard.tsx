"use client";

import type React from "react";

import { useState } from "react";
import type { Job } from "../types/job";
import { MapPin, Calendar, Clock, Bookmark } from "lucide-react";
import { toggleBookmark } from "../services/bookmarks";
import { isAuthenticated } from "../services/auth";

interface JobCardProps {
  job: Job;
  onClick: () => void;
  disabled?: boolean;
  onBookmarkChange?: (jobId: string, isBookmarked: boolean) => void;
}

export default function JobCard({
  job,
  onClick,
  disabled = false,
  onBookmarkChange,
}: JobCardProps) {
  const [isBookmarked, setIsBookmarked] = useState(job.isBookmarked || false);
  const [bookmarkLoading, setBookmarkLoading] = useState(false);
  const [bookmarkError, setBookmarkError] = useState("");

  const handleBookmarkClick = async (e: React.MouseEvent) => {
    e.stopPropagation(); // Prevent card click when clicking bookmark

    if (!isAuthenticated()) {
      setBookmarkError("Please login to bookmark jobs");
      return;
    }

    setBookmarkLoading(true);
    setBookmarkError("");

    try {
      const newBookmarkStatus = await toggleBookmark(job.id, isBookmarked);
      setIsBookmarked(newBookmarkStatus);

      // Notify parent component of bookmark change
      onBookmarkChange?.(job.id, newBookmarkStatus);
    } catch (error) {
      setBookmarkError(
        error instanceof Error ? error.message : "Failed to update bookmark"
      );
      console.error("Bookmark error:", error);
    } finally {
      setBookmarkLoading(false);
    }
  };

  return (
    <div
      className={`bg-white rounded-lg border border-gray-200 p-6 transition-all relative ${
        disabled
          ? "opacity-50 cursor-not-allowed"
          : "hover:shadow-md cursor-pointer hover:border-blue-200"
      }`}
      onClick={disabled ? undefined : onClick}
      data-testid="job-card"
    >
      {/* Bookmark Button */}
      <button
        onClick={handleBookmarkClick}
        disabled={bookmarkLoading}
        className={`absolute top-4 right-4 p-2 rounded-full transition-all ${
          isBookmarked
            ? "bg-blue-100 text-blue-600 hover:bg-blue-200"
            : "bg-gray-100 text-gray-400 hover:bg-gray-200 hover:text-gray-600"
        } ${bookmarkLoading ? "opacity-50 cursor-not-allowed" : ""}`}
        data-testid="bookmark-btn"
        aria-label={isBookmarked ? "Remove from bookmarks" : "Add to bookmarks"}
      >
        <Bookmark
          className={`w-5 h-5 ${isBookmarked ? "fill-current" : ""}`}
          data-testid="bookmark-icon"
        />
      </button>

      {/* Bookmark Error */}
      {bookmarkError && (
        <div className="absolute top-16 right-4 bg-red-50 border border-red-200 rounded-md p-2 text-xs text-red-600 z-10">
          {bookmarkError}
        </div>
      )}

      <div className="flex gap-4 pr-12">
        {/* Avatar */}
        <div className="flex-shrink-0">
          <img
            src={job.logoUrl || "/placeholder.svg?height=48&width=48&text=Logo"}
            alt={job.orgName}
            className="w-12 h-12 rounded-full object-cover border border-gray-200"
            data-testid="company-logo"
            onError={(e) => {
              const target = e.target as HTMLImageElement;
              target.src = "/placeholder.svg?height=48&width=48&text=Logo";
            }}
          />
        </div>

        {/* Content */}
        <div className="flex-1">
          <div className="flex justify-between items-start mb-2">
            <h3
              className="text-xl font-semibold text-gray-900 hover:text-blue-600 transition-colors"
              data-testid="job-title"
            >
              {job.title}
            </h3>
          </div>

          <div className="flex items-center gap-4 text-sm text-gray-600 mb-3">
            <span className="font-medium" data-testid="company-name">
              {job.orgName}
            </span>
            <div className="flex items-center gap-1">
              <MapPin className="w-4 h-4" />
              <span data-testid="job-location">
                {job.location?.join(", ") || "Location not specified"}
              </span>
            </div>
          </div>

          <p
            className="text-gray-700 mb-4 line-clamp-3"
            data-testid="job-description"
          >
            {job.description}
          </p>

          {/* Categories/Tags */}
          <div
            className="flex gap-2 mb-4 flex-wrap"
            data-testid="job-categories"
          >
            {job.categories?.map((category, index) => (
              <span
                key={index}
                className="px-3 py-1 rounded-full text-xs font-medium bg-blue-100 text-blue-800 border border-blue-200"
              >
                {category}
              </span>
            ))}
            {job.opType && (
              <span className="px-3 py-1 rounded-full text-xs font-medium bg-green-100 text-green-800 border border-green-200">
                {job.opType}
              </span>
            )}
          </div>

          {/* Footer */}
          <div className="flex items-center justify-between text-sm text-gray-500">
            <div className="flex items-center gap-4">
              <div className="flex items-center gap-1">
                <Calendar className="w-4 h-4" />
                <span>
                  Posted{" "}
                  {job.datePosted
                    ? new Date(job.datePosted).toLocaleDateString()
                    : "Recently"}
                </span>
              </div>
              {job.deadline && (
                <div className="flex items-center gap-1">
                  <Clock className="w-4 h-4" />
                  <span>
                    Deadline {new Date(job.deadline).toLocaleDateString()}
                  </span>
                </div>
              )}
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}
