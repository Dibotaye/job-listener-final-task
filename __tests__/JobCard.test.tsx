import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import JobCard from "@/components/JobCard";
import * as bookmarksApi from "@/services/bookmarks";

jest.mock("@/services/bookmarks", () => ({
  toggleBookmark: jest.fn(),
}));

describe("JobCard Component", () => {
  const mockJob = {
    id: "65509e9353a7667de6ef5a60", // ✅ match component's expected prop
    title: "Software Engineer",
    orgName: "Tech Corp",
    location: ["Remote"],
    description: "An exciting role",
    logoUrl: "/logo.png",
    datePosted: "2024-07-17T11:09:02.207Z",
    opType: "Full-time",
    isBookmarked: false, // ✅ initial state
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("renders job details correctly", () => {
    render(<JobCard job={mockJob} onClick={jest.fn()} />);

    expect(screen.getByTestId("job-title")).toHaveTextContent(
      "Software Engineer"
    );
    expect(screen.getByTestId("company-name")).toHaveTextContent("Tech Corp");
    expect(screen.getByTestId("job-location")).toHaveTextContent("Remote");
    expect(screen.getByTestId("job-description")).toHaveTextContent(
      "An exciting role"
    );
  });

  test("calls toggleBookmark when bookmark button is clicked", async () => {
    const user = userEvent.setup();
    (bookmarksApi.toggleBookmark as jest.Mock).mockResolvedValue(true);

    render(<JobCard job={mockJob} onClick={jest.fn()} />);

    await user.click(screen.getByTestId("bookmark-btn"));

    expect(bookmarksApi.toggleBookmark).toHaveBeenCalledWith(
      "65509e9353a7667de6ef5a60", // ✅ matches job.id
      false // ✅ matches initial isBookmarked
    );
  });
});
