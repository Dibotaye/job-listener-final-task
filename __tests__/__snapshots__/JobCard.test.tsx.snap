// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`JobCard Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="bg-white rounded-lg border border-gray-200 p-6 transition-all relative hover:shadow-md cursor-pointer hover:border-blue-200"
    data-testid="job-card"
  >
    <button
      aria-label="Add to bookmarks"
      class="absolute top-4 right-4 p-2 rounded-full transition-all bg-gray-100 text-gray-400 hover:bg-gray-200 hover:text-gray-600 "
      data-testid="bookmark-btn"
    >
      <svg
        class="lucide lucide-bookmark w-5 h-5"
        data-testid="bookmark-icon"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m19 21-7-4-7 4V5a2 2 0 0 1 2-2h10a2 2 0 0 1 2 2v16z"
        />
      </svg>
    </button>
    <div
      class="flex gap-4 pr-12"
    >
      <div
        class="flex-shrink-0"
      >
        <img
          alt="Tech Corp"
          class="w-12 h-12 rounded-full object-cover border border-gray-200"
          data-testid="company-logo"
          src="/placeholder.svg?height=48&width=48&text=Logo"
        />
      </div>
      <div
        class="flex-1"
      >
        <div
          class="flex justify-between items-start mb-2"
        >
          <h3
            class="text-xl font-semibold text-gray-900 hover:text-blue-600 transition-colors"
            data-testid="job-title"
          >
            Software Engineer
          </h3>
        </div>
        <div
          class="flex items-center gap-4 text-sm text-gray-600 mb-3"
        >
          <span
            class="font-medium"
            data-testid="company-name"
          >
            Tech Corp
          </span>
          <div
            class="flex items-center gap-1"
          >
            <svg
              class="lucide lucide-map-pin w-4 h-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M20 10c0 4.993-5.539 10.193-7.399 11.799a1 1 0 0 1-1.202 0C9.539 20.193 4 14.993 4 10a8 8 0 0 1 16 0"
              />
              <circle
                cx="12"
                cy="10"
                r="3"
              />
            </svg>
            <span
              data-testid="job-location"
            >
              Remote
            </span>
          </div>
        </div>
        <p
          class="text-gray-700 mb-4 line-clamp-3"
          data-testid="job-description"
        >
          An exciting role
        </p>
        <div
          class="flex gap-2 mb-4 flex-wrap"
          data-testid="job-categories"
        >
          <span
            class="px-3 py-1 rounded-full text-xs font-medium bg-blue-100 text-blue-800 border border-blue-200"
          >
            Engineering
          </span>
          <span
            class="px-3 py-1 rounded-full text-xs font-medium bg-green-100 text-green-800 border border-green-200"
          >
            Full-time
          </span>
        </div>
        <div
          class="flex items-center justify-between text-sm text-gray-500"
        >
          <div
            class="flex items-center gap-4"
          >
            <div
              class="flex items-center gap-1"
            >
              <svg
                class="lucide lucide-calendar w-4 h-4"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 2v4"
                />
                <path
                  d="M16 2v4"
                />
                <rect
                  height="18"
                  rx="2"
                  width="18"
                  x="3"
                  y="4"
                />
                <path
                  d="M3 10h18"
                />
              </svg>
              <span>
                Posted 8/12/2025
              </span>
            </div>
            <div
              class="flex items-center gap-1"
            >
              <svg
                class="lucide lucide-clock w-4 h-4"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="12"
                  cy="12"
                  r="10"
                />
                <polyline
                  points="12 6 12 12 16 14"
                />
              </svg>
              <span>
                Deadline 9/1/2025
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
